compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1031,11 +1056,34 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
                     //    must be instantiated, making the two tycons equal
                     val tyconIsInjective =
                       (tycon1sym.isClass || tycon2sym.isClass)
-                        && (if touchedGADTs then gadtIsInstantiated else true)
-                    def checkSubArgs() = isSubArgs(args1, args2, tp1, tparams)
-                    // we only record GADT constraints if *both* tycons are effectively injective
-                    if (tyconIsInjective) checkSubArgs()
-                    else inFrozenGadt { checkSubArgs() }
+                      && (!touchedGADTs || gadtIsInstantiated)
+
+                    def checkArgs(): Boolean =
+                      if tycon1sym == tycon2sym && tycon1sym.isAliasType then
+                        // Either compare arguments or compare dealiased types.
+                        // Note: This test should be done with an `either`, but this fails
+                        // for hk-alias-unification.scala The problem is not GADT boundschecking;
+                        // that is taken care of by the outer inFrozenGadtIf test. The problem is
+                        // that hk-alias-unification.scala relies on the right isSubArgs being performed
+                        // when constraining the result type of a method. But since #8635 we are only allowed
+                        // to use a necessaryEither, since we might otherwise constrain too much.
+                        // So strictly speaking by going to a sufficientEither we overshoot and narrow
+                        // the constraint unnecessarily. But unlike in the situation of #8365 with test cases
+                        // and-inf.scala and or-inf.scala, we don't cut off solutions in principle by dealiasing.
+                        // We "just" risk missing an opportunity to do a desired hk-type inference. This
+                        // is hopefully a less frequent corner case. Where it goes wrong compared to the status
+                        // before this fix: If isSubArgs succeeds and then the dealised test also succeeds
+                        // while constraining strictly less than the isSubArgs, then the we used to pick the
+                        // constraint after isSubArgs, but we now pick the constraint after the dealiased test.
+                        // There's one test in shapeless/deriving that had to be disabled
+                        // (modules/deriving/src/test/scala/shapeless3/deriving/deriving.scala, value v7 in the functor test).
```

Comment:
FYI @milessabin, how important is that particular test?

Reply:
That's an important test. It demonstrates the ability to derive functor instances for nested type constructors.

It would at least be good to know why `v7` fails whereas `v6` succeeds. Both involve recursion. The only difference is that the former failing one has an additional non-recursive wrapper, which is itself successfully derived in `v3`.